flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImpl.java

```diff
@@ -0,0 +1,221 @@
+		return inProgressPartitionTable.hasTrackedPartitions(jobId) || finishedPartitionTable.hasTrackedPartitions(jobId);
+	}
+
+	@Override
+	public void markJobActive(JobID jobId) {
+		activeJobs.add(jobId);
+	}
+
+	@Override
+	public void releaseFinishedPartitions(JobID jobId, Collection<ResultPartitionID> resultPartitionIds) {
+		// maybe double-check that all partitions are finished?
+		finishedPartitionTable.stopTrackingPartitions(jobId, resultPartitionIds);
+		backingShuffleEnvironment.releasePartitions(resultPartitionIds);
+	}
+
+	@Override
+	public void releaseAllFinishedPartitionsAndMarkJobInactive(JobID jobId) {
+		activeJobs.remove(jobId);
+		Collection<ResultPartitionID> finishedPartitionsForJob = finishedPartitionTable.stopTrackingPartitions(jobId);
+		backingShuffleEnvironment.releasePartitions(finishedPartitionsForJob);
+	}
+
+	/**
+	 * This method wraps partition writers if the partition is not relased and consumption
+	 * and introduces callbacks into the lifecycle methods of the {@link ResultPartitionWriter}.
+	 */
+	@Override
+	public Collection<ResultPartitionWriter> createResultPartitionWriters(
+		JobID jobId,
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup) {
+
+		final Collection<? extends ResultPartitionWriter> resultPartitionWriters = backingShuffleEnvironment
+			.createResultPartitionWriters(taskName, executionAttemptID, resultPartitionDeploymentDescriptors, outputGroup, buffersGroup);
+
+		// determine which partitions are not released on consumption, and whose writers need to be wrapped
+		final Set<ResultPartitionID> externallyManagedPartitions = resultPartitionDeploymentDescriptors.stream()
+			.filter(descriptor -> !descriptor.isReleasedOnConsumption())
+			.map(ResultPartitionDeploymentDescriptor::getShuffleDescriptor)
+			.map(ShuffleDescriptor::getResultPartitionID)
+			.collect(Collectors.toSet());
+
+		// wrap partition writers to introduce callbacks for lifecycle events
+		final Collection<ResultPartitionWriter> wrappedResultPartitionWriters = new ArrayList<>(resultPartitionWriters.size());
+		for (ResultPartitionWriter resultPartitionWriter : resultPartitionWriters) {
+			final ResultPartitionWriter maybeWrappedWriter;
+			if (externallyManagedPartitions.contains(resultPartitionWriter.getPartitionId())) {
+				maybeWrappedWriter = new NotifyingResultPartitionWriter(
+					resultPartitionWriter,
+					partitionId -> notifyPartitionSetup(jobId, partitionId),
+					partitionId -> notifyPartitionFailed(jobId, partitionId),
+					partitionId -> notifyPartitionFinished(jobId, partitionId));
+			} else {
+				maybeWrappedWriter = resultPartitionWriter;
+			}
+			wrappedResultPartitionWriters.add(maybeWrappedWriter);
+		}
+
+		return wrappedResultPartitionWriters;
+	}
+
+	private void notifyPartitionSetup(JobID jobId, ResultPartitionID resultPartitionId) {
```

Comment:
The task life-cycle only guarantees a finish()/fail() call after setup().

If we add them before (in the Task constructor) we need to handle additional (super-rare) edge-cases in the TaskExecutor in a way that is disconnected from the actual partition life-cycle.

If we want to go down this route we don't need the wrappers at all, as we might just as well handle everything on the TaskExecutor level. (add partitions on submission, remove partitions in `unregisterTaskAndNotifyFinalState`)

Reply:
To expand a bit further on this, because this is quite interesting and one of the core issues that had to be handled:
let's say we register partitions in `createPartitionWriters`, which is the most natural way to do it while we have wrappers.
For the normal cases where partitions are finished/failed the environment can be kept in the loop about it through listeners.
For cases where tasks fail before the partition was setup (i.e., anywhere before the main try-catch block in run()) this is no longer possible; there is no callback that the environment could rely on, because the task won't interact with the environment any further. It just exits and _usually_ informs the TE about the failure.

So, who updates the tracking now?

We would either have the task or TE un-register the partitions. We don't want the task to do it, as this isn't something it deals with anywhere else.
So we'd have the TE do it, but this basically leaks implementation details of the task/JobAwareShuffleEnvironment; after all it's just a crutch to handle some cases these components aren't handling.

One could argue that we could avoid this being a special case by _always_ having the TM update the tracking after the task shuts down. This approach is _really_ attractive because you no longer need wrappers, but you're kind of flying blind; the tracking of the partitions no longer has any real relation to what actually happens in the task/shuffle environment. This also makes it more difficult to understand the code; the shuffle environment and TE both do their own thing that somehow is in sync (hopefully).

That said, the wrappers only partially solve the relation issue; after all the shuffle environment could be releasing partitions left and right at will and we would be none the wiser. part of the `ResultPartitionWriter` interface.

Both approaches have their flaws.

Reply:
I guess the answer would be that we need a proper lifecycle management of the `Task`. If the task fails before entering the `try-catch` block, then either it needs to make sure that all of its resources get released or the contract is that it tells the owner (`TaskExecutor`) and the owner calls a releaseResources methods where everything is cleaned up.

Reply:
From the perspective of the task there is nothing to clean up. The task itself only allocates something within the try-catch block; that's why it doesn't need to interact with the shuffle environment in any way if it fails there.
The task constructor specifically says that nothing requiring cleanup should be done in the constructor, which we would do if we were to hook into `createResultPartitionWriters`.